### PR TITLE
Issue/11

### PR DESCRIPTION
**Pull request by 0x1NotMe**
## Description

Add initialize function to vault
Removed constructor arguments
Add support for two OpenZeppelin library versions 

Fixes # (issue)
- maxAvailableShares was using strategies rather than withdrawalQueue when iterating over strategies

## Checklist

- [x] I have run the tests on my machine
- [x] I have followed commitlint guidelines
- [x] I have rebased my changes to the latest version of the main branch

### CODE REVIEW

File: contracts/Vault.sol [hunk 1]
```diff
@@ -134,7 +191,7 @@ contract Vault {
 
     function setRewards(address rewards_) external {
         require(msg.sender == governance, "This may only be called by governance.");
-        assert(rewards_ != address(0) || rewards_ != _self);
+        assert(rewards_ != address(0) || rewards_ != address(this));
```
Comment (bitcoinbrisbane) [inline contracts/Vault.sol:194]: why remove `_self`?

File: contracts/Vault.sol [hunk 2]
```diff
@@ -87,27 +81,90 @@ contract Vault {
...
+        address governance_,
+        address rewards_,
+        string memory nameOverride_,
+        string memory symbolOverride_,
+        address guardian_,
+        address management_,
+        address healthCheck_
+    ) external {
+
+        require(activation == 0, 'already initilized');
```
Comment (bitcoinbrisbane) [inline contracts/Vault.sol:115]:
```suggestion
        require(activation == 0, "already initilized");
```

File: contracts/Vault.sol [hunk 3]
```diff
@@ -72,8 +65,9 @@ contract Vault {
     address private pendingGovernance;
     address private healthCheck;
 
-    address private immutable _self;
-
+    //TODO verfify this is not required
+    //address private immutable _self;
```
Comment (bitcoinbrisbane) [inline contracts/Vault.sol:69]: Lets roll this back.  It saves gas by caching address(this)

File: contracts/Vault.sol (hunk 1 above)
Comment (0x1NotMe) [reply contracts/Vault.sol]: done

File: contracts/Vault.sol (hunk 2 above)
Comment (0x1NotMe) [reply contracts/Vault.sol]: done

File: contracts/Vault.sol (hunk 3 above)
Comment (0x1NotMe) [reply contracts/Vault.sol]: done

File: contracts/Vault.sol [hunk 4]
```diff
@@ -44,15 +43,17 @@ struct StrategyParams {
     uint256 totalLoss;
 }
 
-// TODO: Implements ERC20
 contract Vault is ERC20 {
     using SafeERC20 for IERC20;
 
     IERC20 public immutable token;
```
Comment (LichuAcu) [inline contracts/Vault.sol:49]: If token is immutable, it cannot be changed in _initialize_ function (line 116)

File: contracts/Vault.sol [hunk 5]
```diff
@@ -80,27 +81,90 @@ contract Vault is ERC20 {
...
+        emit UpdateManagement(management);
+        rewards = rewards_;
+        emit UpdateRewards(rewards);
+        guardian = guardian_;
+        emit UpdateGuardian(guardian);
+        performanceFee = 1000;  // 10% of yield (per Strategy)
+        emit UpdatePerformanceFee(uint256(1000));
+        managementFee = 200;  // 2% per year
+        emit UpdateManagementFee(uint256(200));
+        healthCheck = healthCheck_;
```
Comment (LichuAcu) [inline contracts/Vault.sol:152]:
```suggestion
        healthCheck = healthCheck_;
        emit UpdateHealthCheck(healthCheck);
```

File: contracts/Vault.sol (hunk 4 above)
Comment (LichuAcu) [reply contracts/Vault.sol:49]:
```suggestion
    IERC20 public token;
```

**Review by LichuAcu — COMMENTED** [review]
Added a missing event and changed _token_ variable declaration.

**Review by bitcoinbrisbane — APPROVED** [review]

**Pull request merged**